File: 백준/Gold/15683. 감시/감시.cc
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <cstring>

using namespace std;

typedef struct
{
	int x;
	int y;
	int type;
	int direction;//0 상 1 우 2 하 3 좌
}CCTV;

int N, M;
int board[8][8];
vector<CCTV> cctvs;

int dx[] = { 0,1,0,-1 };
int dy[] = { -1,0,1,0 };

void make(CCTV& cctv)
{
	vector<int> dirs;
	switch (cctv.type)
	{
	case 1:
		dirs.push_back(cctv.direction);
		break;
	case 2:
		dirs.push_back(cctv.direction);//보는방향, 반대방향
		dirs.push_back((cctv.direction + 2+4) % 4);
		break;
	case 3:
		dirs.push_back(cctv.direction);//보는장향, 오른쪽으로 90도 꺾은 방향
		dirs.push_back((cctv.direction + 1+4) % 4);
		break;
	case 4:
		dirs.push_back(cctv.direction);//보는방향, 좌, 우로 90도 꺾은 방향
		dirs.push_back((cctv.direction + 1 + 4) % 4);
		dirs.push_back((cctv.direction - 1 + 4) % 4);
		break;
	case 5:
		dirs.push_back(0);//그냥 다
		dirs.push_back(1);
		dirs.push_back(2);
		dirs.push_back(3);
		break;
	default:
		break;
	}
	for (auto& dir : dirs)
	{
		int nextx = cctv.x + dx[dir];
		int nexty = cctv.y + dy[dir];
		while (true)
		{
			if (nextx < 0 || nextx >= M || nexty < 0 || nexty >= N || board[nexty][nextx] == 6)
				break;

			if (board[nexty][nextx] == 0)
			{
				board[nexty][nextx] = 8;
			}
			nextx = nextx + dx[dir];
			nexty = nexty + dy[dir];
		}
	}
}

int minsafezone = 64;
int boardsave[8][8];

void dfs(int index)
{
	if (index == cctvs.size())
	{
		//여기서 빈칸 확인하고 빈칸 최소 되는거 구하기
		memcpy(boardsave, board, sizeof(board));
		for (auto& cctv : cctvs)
		{
			make(cctv);
		}
		int safezone = 0;
		for (int y = 0; y < N; y++)
		{
			for (int x = 0; x < M; x++)
			{
				if (board[y][x] == 0)
					safezone++;
			}
		}

		minsafezone = min(minsafezone, safezone);
		memcpy(board,boardsave, sizeof(board));
		return;
	}

	//방향 전부 돌리면서 dfs 호출
	cctvs[index].direction = 0;
	dfs(index + 1);

	cctvs[index].direction = 1;
	dfs(index + 1);

	cctvs[index].direction = 2;
	dfs(index + 1);

	cctvs[index].direction = 3;
	dfs(index + 1);
}

int main()
{
	cin >> N >> M;

	int cctvnum = 0;
	for (int y = 0; y < N; y++)
	{
		for (int x = 0; x < M; x++)
		{
			cin >> board[y][x];
			if (board[y][x] != 0 && board[y][x] != 6)
				cctvs.push_back({ x,y,board[y][x] ,0});
		}
	}
	dfs(0);
	cout << minsafezone << endl;
}

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

class CCTV{
	public int x,y,type,dir;
}

public class Main {
	static int N, M;
	static int map[][];
	static int savemap[][];
	static CCTV cctvs[] = new CCTV[8];
	static int cctvnum = 0;
	static int dx[] = {-1,0,1,0};
	static int dy[] = {0,-1,0,1};
	static int Solution = Integer.MAX_VALUE;
	
	public static void EraseMarkedPart(CCTV cctv) {
		ArrayList <Integer> dirs = new ArrayList<>();
		
		switch(cctv.type) {
		case 1:
			dirs.add(cctv.dir);
			break;
		case 2:
			dirs.add(cctv.dir);
			dirs.add((cctv.dir+ 2) % 4);
			break;
		case 3:
			dirs.add(cctv.dir);
			dirs.add((cctv.dir + 1) % 4);
			break;
		case 4:
			dirs.add(cctv.dir);
			dirs.add((cctv.dir + 1) % 4);
			dirs.add((cctv.dir + 2) % 4);
			break;
		case 5:
			dirs.add(cctv.dir);
			dirs.add((cctv.dir + 1) % 4);
			dirs.add((cctv.dir + 2) % 4);
			dirs.add((cctv.dir + 3) % 4);
			break;
		}
		for (int dir : dirs) {
			int currx = cctv.x;
			int curry = cctv.y;
			
			while(true) {
				int nextx = currx + dx[dir];
				int nexty = curry + dy[dir];
				
				if (nextx <0 || nextx >= M || nexty <0 || nexty >= N)
					break;
				if (map[nexty][nextx] == 6)
					break;
				
				if (map[nexty][nextx] < 1)
					map[nexty][nextx] = 9;
				currx = nextx;
				curry = nexty;
			}
			
		}
	}
	
	public static void dfs(int index) {
		if(index == cctvnum) {
			for(int y= 0; y< N;y++) {
				for(int x = 0; x< M; x++) {
					savemap[y][x] = map[y][x];
				}
			}
			
			for(int i =0 ; i < cctvnum;i++) 
				EraseMarkedPart(cctvs[i]);
			

			int unmarkednum = 0;
			for(int y= 0;  y< N; y++) {
				for(int x = 0; x < M ;x++) {
					if (map[y][x] == 0) {
						unmarkednum++;
					}
				}
			}
			Solution = Math.min(unmarkednum, Solution);
			
			for(int y= 0; y< N;y++) {
				for(int x = 0; x< M; x++) {
					map[y][x] =savemap[y][x];
				}
			}
			return;
		}
		
		for(int d =0; d<4;d++) {
			cctvs[index].dir = d;
			dfs(index+1);
		}
	}
	
	public static void main(String args[]) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer (br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		map = new int[N][M];
		savemap = new int[N][M];
		
		for(int y= 0;  y< N; y++) {
			st = new StringTokenizer (br.readLine());
			for(int x = 0; x < M ;x++) {
				map[y][x] = Integer.parseInt(st.nextToken());
				if (map[y][x] != 0 &&map[y][x] != 6 ) {
					cctvs[cctvnum] = new CCTV();
					cctvs[cctvnum].x = x;
					cctvs[cctvnum].y = y;
					cctvs[cctvnum].type = map[y][x];
					
					cctvnum++;
				}
			}
		}
		
		
		
		dfs(0);
		System.out.println(Solution);
	}
}